home/vasco-santos/2021-03-exploration-report-1-on-js-lotus-client.md

```diff
@@ -0,0 +1,102 @@
+# Experiment with lotus client
+
+## Deal Flow
+
+Requisites:
+- A wallet id with FIL
+
+Flow:
+1. import file
+  * `provider.importFile(data: String): Promise<CID>`
+2. fetch file info for deal
+  * `client.clientDealSize(cid: Cid): Promise<DataSize>`
+3. list miners
+  * `client.stateListMiners(tipSetKey: Cid[]): Promise<Array<string>>`
+4. get miner info for each miner
+  * `client.stateMinerInfo(address: string, tipSetKey: Cid[]): Promise<MinerInfo>`
+5. find the price and conditions of each miner
+  * `client.clientQueryAsk(id: string, address: string): Promise<StorageAsk>`
+6. start a deal with the intended miners
+  * `clientStartDeal (startDealParams: StartDealParams): Promise<Cid>`
+
+
+```js
+// const wallet = ...
+const cidString = await provider.importFile('this is random text!')
+
+const cid = {
+  '/': cidString
+}
+
+const dataSize = await client.clientDealSize(cid)
+
+const head = await client.chainHead()
+const miners = await client.stateListMiners(head.Cids)
```

Comment:
oh, this is interesting! so you query the state tree at a particular head. I was recently interested in looking at some historical data to calculate some diffs between a particular point in history and _now_ but thought that I'd have to hack lotus to do it manually. But if this pattern is consistent across the RPC then it could be quite useful if a node has enough historical data in it!

Reply:
You can provide a particular head, but it is not required (unless you are using typescript like I was 😓 ). Current ts definitions require the parameter, but the CLI does not require the head, so I believe that the remote client does not require the head, but as you mentioned it is nice to have